### PR TITLE
improve usability of schnellru::LruMap and LruCache (LinkedHashSet cache) 

### PR DESCRIPTION
**Pull request by emhane**
- Wrapper of schnellru::LruMap that implements Debug to make it easier to include in structs for which Debug is derived.
- Configurable eviction feedback for LruCache (LinkedHashSet cache).

### CODE REVIEW

File: crates/net/network/src/cache.rs [hunk 1]
```diff
@@ -9,12 +19,19 @@ use std::{borrow::Borrow, hash::Hash, num::NonZeroUsize};
 pub struct LruCache<T: Hash + Eq> {
     limit: NonZeroUsize,
     inner: LinkedHashSet<T>,
+    tx: Option<UnboundedSender<T>>,
```
Comment (Rjected) [inline crates/net/network/src/cache.rs:22]: This is cool, but I would add some docs for this and change the name from `tx` to something that makes it more clear that this is the eviction feedback sender

File: crates/net/network/src/cache.rs [hunk 2]
```diff
@@ -9,12 +19,19 @@ use std::{borrow::Borrow, hash::Hash, num::NonZeroUsize};
...
-    /// Creates a new `LruCache` using the given limit
+    /// Creates a new [`LruCache`] using the given limit
     pub fn new(limit: NonZeroUsize) -> Self {
-        Self { inner: LinkedHashSet::new(), limit }
+        Self { inner: LinkedHashSet::new(), limit, tx: None }
+    }
+
+    /// Creates a new [`LruCache`] using the given limit and with a channel for sending eviction
+    /// feedback. Note, that it is up to the caller's context to drain the buffer!
+    pub fn new_with_feedback(limit: NonZeroUsize, tx: UnboundedSender<T>) -> Self {
```
Comment (Rjected) [inline crates/net/network/src/cache.rs:33]: looks like `new_with_feedback` and `remove` will need to be `pub(crate)` to satisfy clippy

File: crates/net/network/src/cache.rs [hunk 3]
```diff
@@ -1,5 +1,15 @@
+use derive_more::{Deref, DerefMut};
 use linked_hash_set::LinkedHashSet;
-use std::{borrow::Borrow, hash::Hash, num::NonZeroUsize};
+use schnellru::{self, ByLength};
```
Comment (Rjected) [inline crates/net/network/src/cache.rs:3]:
```suggestion
use schnellru::ByLength;
```

**Review by Rjected — DISMISSED** [review]
just a nit about documentation, and I think schnellru needs to be added as a dependency to the network crate, otherwise this looks good to me

File: crates/net/network/src/cache.rs (hunk 1 above)
Comment (Rjected) [reply crates/net/network/src/cache.rs:22]: also some docs for the `LruCache` struct about this behavior

Comment (mattsse) [inline crates/net/network/src/cache.rs:22]: I would prefer this feature in a new type.
The interface of the LruCache is very small, so I think a wrapper that that has

```rust
struct LruCache(Sender|Emitter|...) {
  inner: LruCache,
  tx: UnboundedSender
}
```

would be a bit cleaner imo

File: crates/net/network/src/cache.rs [hunk 4]
```diff
@@ -70,8 +102,48 @@ where
     }
 }
 
+/// Wrapper of [`schnellru::LruMap`] that implements [`fmt::Debug`].
+#[derive(Deref, DerefMut)]
+pub struct LruMap<K, V>(schnellru::LruMap<K, V>)
```
Comment (mattsse) [inline crates/net/network/src/cache.rs:107]: ah right the schnellru map does indeed not impl debug...

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/network/src/cache.rs (hunk 3 above)
Comment (emhane) [reply crates/net/network/src/cache.rs:3]: I'm using `schnellru::LruMap` to be able to make new type with same name `LruMap`, so I need to bring the whole module into scope. I forgot to add schnellru crate to the manifest though when I moved this code out to its own branch yesterday that's why clippy says the external crate doesn't exist.

File: crates/net/network/src/cache.rs (hunk 1 above)
Comment (emhane) [reply crates/net/network/src/cache.rs:22]: valid, will make a new type though https://github.com/paradigmxyz/reth/pull/6016#discussion_r1448456693

**Comment by emhane** [comment]
came to the same conclusion as mattsse https://github.com/paradigmxyz/reth/pull/6012#discussion_r1449018050 @mattsse @Rjected

File: crates/net/network/src/cache.rs [hunk 5]
```diff
@@ -36,14 +44,33 @@ impl<T: Hash + Eq> LruCache<T> {
         false
     }
 
+    /// Same as [`Self::insert`] but returns a tuple, where the second index is the evicted value,
+    /// if one was evicted.
+    pub fn _insert_with_eviction_feedback(&mut self, entry: T) -> (bool, Option<T>) {
```
Comment (mattsse) [inline crates/net/network/src/cache.rs:49]: I see, we can't express all 3 cases with a single option

* inserted new + not evicted
* inserted (already existing)
* inserted new + evicted

so this tuple makes sense.

I'd like to drop the _ here too.

I thought about a different name for a minute but I can't come up with anything.
but perhaps `insert_and_get_evicted`? fine with this though

Comment (mattsse) [inline crates/net/network/src/cache.rs:49]: can we reuse this function in:
`insert{self.insert_evicttion().map(|(inserted,_)| inserted)}`

File: crates/net/network/src/cache.rs [hunk 6]
```diff
@@ -36,14 +44,33 @@ impl<T: Hash + Eq> LruCache<T> {
...
-    /// If the `LruCache` is empty this will return None.
+    /// If the `LruCache` is empty or if the eviction feedback is
+    /// configured, this will return None.
     #[inline]
     fn remove_lru(&mut self) -> Option<T> {
         self.inner.pop_front()
     }
 
+    /// Expels the given value. Returns true if the value existed.
+    pub fn _remove(&mut self, value: &T) -> bool {
```
Comment (mattsse) [inline crates/net/network/src/cache.rs:70]: I think we can drop the _ prefix here, no need to hide it

File: crates/net/network/src/cache.rs [hunk 7]
```diff
@@ -36,14 +44,33 @@ impl<T: Hash + Eq> LruCache<T> {
...
+            return (true, None)
+        }
+        (false, None)
+    }
+
     /// Remove the least recently used entry and return it.
     ///
-    /// If the `LruCache` is empty this will return None.
+    /// If the `LruCache` is empty or if the eviction feedback is
+    /// configured, this will return None.
```
Comment (mattsse) [inline crates/net/network/src/cache.rs:63]: > if the eviction feedback is configured,

no longer relevant, right?

**Review by mattsse — CHANGES_REQUESTED** [review]
a few smol suggestions.

lgtm otherwise

File: crates/net/network/src/cache.rs (hunk 6 above)
Comment (emhane) [reply crates/net/network/src/cache.rs:70]: tell that to clippy, jk

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**